File: StudentGrade/python/student_grade.py
student_quiz_results = []

def set_student_quiz_results(scores):
    global student_quiz_results
    student_quiz_results = scores

def get_total_scores():
    student_total_scores = []

    for student_quiz_result in student_quiz_results:
        total = 0

        for score in student_quiz_result:
            total += score

        student_total_scores.append(total)

    return student_total_scores

def get_student_position():

    position = []

    student_total_scores = get_total_scores()

    sorted_student_total_scores = sorted(student_total_scores, reverse=True)

    for total_score in student_total_scores:

        index = sorted_student_total_scores.index(total_score)

        position.append(index + 1)

    return position


def get_summary_for_subject(subject_number):

    best_student_number = 0
    best_student_score = student_quiz_results[0][subject_number]

    worst_student_number = 0
    worst_student_score = student_quiz_results[0][subject_number]

    total_score = 0
    fail_count = 0
    pass_count = 0

    number_of_students = len(student_quiz_results)

    for student_index in range(number_of_students):

        score = student_quiz_results[student_index][subject_number]

        total_score += score

        if score < 50:
            fail_count += 1
        else:
            pass_count += 1

        if score > best_student_score:
            best_student_score = score
            best_student_number = student_index

        if score < worst_student_score:
            worst_student_score = score
            worst_student_number = student_index

    average = total_score / number_of_students

    return f"""Subject {subject_number + 1}
Highest scoring student is: Student {best_student_number + 1} scoring {best_student_score}
Lowest scoring student is: Student {worst_student_number + 1} scoring {worst_student_score}
Total score is: {total_score}
Average score is: {average:.2f}
Number of passes: {pass_count}
Number of fails: {fail_count}""";


def get_difficulty_summary_for_all_subject():
    hardest_subject_index = 0
    highest_fail_count = 0

    easiest_subject_index = 0
    highest_pass_count = 0

    number_of_students = len(student_quiz_results)

    number_of_subjects = len(student_quiz_results[0])

    for subject_index in range(number_of_subjects):

        fail_count = 0

        pass_count = 0

        for student_index in range(number_of_students):

            score = student_quiz_results[student_index][subject_index]

            if score < 50:
                fail_count += 1
            else:
                pass_count += 1

        if fail_count > highest_fail_count:

            highest_fail_count = fail_count

            hardest_subject_index = subject_index

        if pass_count > highest_pass_count:

            highest_pass_count = pass_count

            easiest_subject_index = subject_index

    return f"The hardest subject is Subject {hardest_subject_index + 1} with {highest_fail_count} failures\nThe easiest subject is Subject {easiest_subject_index + 1} with {highest_pass_count} passes"

def get_summary_for_overall_score():

    best_student_number = 0
    best_subject_number = 0
    best_student_score = student_quiz_results[0][0]

    worst_student_number = 0
    worst_subject_number = 0
    worst_student_score = student_quiz_results[0][0]

    number_of_students = len(student_quiz_results)

    number_of_subjects = len(student_quiz_results[0])

    for student_index in range(number_of_students):

        for subject_index in range(number_of_subjects):

            score = student_quiz_results[student_index][subject_index]

            if score > best_student_score:

                best_student_score = score

                best_student_number = student_index

                best_subject_number = subject_index

            if score < worst_student_score:

                worst_student_score = score

                worst_student_number = student_index

                worst_subject_number = subject_index

    return f"The overall Highest score is scored by Student {best_student_number + 1} in subject {best_subject_number + 1} scoring {best_student_score}\nThe overall Lowest score is scored by Student {worst_student_number + 1} in subject {worst_subject_number + 1} scoring {worst_student_score}"

def get_class_summary():

    student_total_scores = get_total_scores()

    number_of_students = len(student_quiz_results)

    position = get_student_position()

    worst_position = max(position)

    best_student_index = position.index(1)

    worst_student_index = position.index(worst_position)

    best_student_score = student_total_scores[best_student_index]

    worst_student_score = student_total_scores[worst_student_index]

    return [[best_student_index + 1, best_student_score], [worst_student_index + 1, worst_student_score]]


def get_total_subject_scores():

    student_total_scores = get_total_scores()

    total_score = 0

    for score in student_total_scores:

        total_score += score

    return total_score

def display_student_result_table():

    position = get_student_position()

    number_of_students = len(student_quiz_results)

    number_of_subjects = len(student_quiz_results[0])

    print("*" * 60)

    print(f"{'STUDENT':<12}", end="")

    for index in range(number_of_subjects):
        print(f"{f'SUB{index + 1}':>6}", end="")

    print(f"{'TOT':>6}{'AVE':>6}{'POS':>8}")

    print("*" * 60)

    for index in range(number_of_students):
        print(f"{f'Student {index + 1}':<10}", end="")

        total = 0

        for score_index in range(number_of_subjects):
            score = student_quiz_results[index][score_index]
            print(f"{score:>6}", end="")
            total += score

        score_average = total / number_of_subjects

        print(f"{total:>8}{score_average:>8.2f}{position[index]:>4}")

    print("*" * 60)

def display_subject_summary():

    number_of_subjects = len(student_quiz_results[0])

    print("*" * 60)

    print("SUBJECT SUMMARY")

    for subject_number in range(number_of_subjects):

        subject_summary = get_summary_for_subject(subject_number)

        print(subject_summary, end="\n\n")

    difficult_summary = get_difficulty_summary_for_all_subject()

    overall_score_summary = get_summary_for_overall_score()

    print(difficult_summary)

    print(overall_score_summary)

    print("*" * 60)

def display_class_summary():

    class_summary = get_class_summary()

    class_total_score = get_total_subject_scores()

    average = class_total_score / len(student_quiz_results)

    print("CLASS SUMMARY")

    print("*" * 60)

    print(f"Best Graduating Student is: Student {class_summary[0][0]} scoring {class_summary[0][1]}")

    print("*" * 60)

    print()

    print("!" * 60)

    print(f"Worst Graduating Student is: Student {class_summary[1][0]} scoring {class_summary[1][1]}")

    print("!" * 60)

    print()

    print("*" * 60)

    print(f"Class total score is: {class_total_score}")

    print(f"Class Average score is: {average:.2f}")

    print("*" * 60)

File: StudentGrade/java/StudentGrade.java
import java.util.Arrays;

public class StudentGrade {

    int[][] studentsResults;

    int studentCount;

    int subjectCount;

    public StudentGrade(int[][] studentsQuizResults) {

        studentsResults = studentsQuizResults;

        studentCount = studentsQuizResults.length;

        subjectCount = studentsQuizResults[0].length;
    }

    public int[] getStudentTotalScore() {
        int[] studentTotalScore = new int[studentCount];

        for (int studentIndex = 0; studentIndex < studentCount; studentIndex++) {

            int totalScore = 0;

            for (int subjectIndex = 0; subjectIndex < subjectCount; subjectIndex++) {

                totalScore += studentsResults[studentIndex][subjectIndex];
            }

            studentTotalScore[studentIndex] = totalScore;
        }

        return studentTotalScore;
    }

    public int[] getPosition() {
        int[] position = new int[studentCount];

        int[] studentTotalScore = getStudentTotalScore();

        int[] sortedStudentTotalScore = Arrays.copyOf(studentTotalScore, studentCount);

        Arrays.sort(sortedStudentTotalScore);

        for (int index = 0;  index < studentCount; index++) {

            for (int innerIndex = 0; innerIndex < studentCount; innerIndex++) {

                if (studentTotalScore[index] == sortedStudentTotalScore[innerIndex]) {

                    position[index] = studentCount - innerIndex;

                    break;
                }
            }
        }

        return position;
    }

    public String getSummaryForSubject(int subjectNumber) {
        int highestStudent = 0;
        int highestScore = studentsResults[0][subjectNumber];

        int lowestStudent = 0;
        int lowestScore = studentsResults[0][subjectNumber];

        int totalScore = 0;
        double averageScore;

        int passesCount = 0;
        int failCount = 0;

        for (int index = 0; index < studentCount; index++) {
            int score = studentsResults[index][subjectNumber];
            totalScore += score;

            if (score < 50)
                failCount++;
            else
                passesCount++;

            if (score > highestScore) {
                highestScore = score;
                highestStudent = index;
            }

            if (score < lowestScore) {
                lowestScore = score;
                lowestStudent = index;
            }
        }

        averageScore = (double) totalScore / studentCount;

        return String.format("""
            Subject %d
            Highest scoring student is: Student %d scoring %d
            Lowest scoring student is: Student %d scoring %d
            Total score is: %d
            Average score is: %.2f
            Number of passes: %d
            Number of fails: %d
            """,
            ++subjectNumber,
            ++highestStudent, highestScore,
            ++lowestStudent, lowestScore,
            totalScore,
            averageScore,
            passesCount,
            failCount);
    }

    public int[][] getSubjectDifficultyDetails() {
        int[][] subjectDifficultyDetails = new int[2][2];

        int hardestSubjectNumber = 0;
        int highestFailureCount = 0;

        int easiestSubjectNumber = 0;
        int highestSuccessCount = 0;

        for (int subjectIndex = 0; subjectIndex < subjectCount; subjectIndex++) {

            int failCount = 0;

            int passCount = 0;

            for (int studentIndex = 0; studentIndex < studentCount; studentIndex++) {

                int score = studentsResults[studentIndex][subjectIndex];

                if (score < 50)
                    failCount++;
                else
                    passCount++;
            }

            if (failCount > highestFailureCount) {
                highestFailureCount = failCount;
                hardestSubjectNumber = subjectIndex;
            }

            if (passCount > highestSuccessCount) {
                highestSuccessCount = passCount;
                easiestSubjectNumber = subjectIndex;
            }
        }


        subjectDifficultyDetails[0][0] = ++hardestSubjectNumber;
        subjectDifficultyDetails[0][1] = highestFailureCount;

        subjectDifficultyDetails[1][0] = ++easiestSubjectNumber;
        subjectDifficultyDetails[1][1] = highestSuccessCount;

        return subjectDifficultyDetails;
    }

    public int[][] getOverallScoreDetails() {

        int[][] overallScoreDetails = new int[2][3];

        int bestStudentNumber = 0;
        int bestSubjectNumber = 0;
        int bestStudentScore = studentsResults[0][0];

        int worstStudentNumber = 0;
        int worstSubjectNumber = 0;
        int worstStudentScore = studentsResults[0][0];

        for (int studentIndex = 0; studentIndex < studentCount; studentIndex++) {

            for (int subjectIndex = 0; subjectIndex < subjectCount; subjectIndex++) {

                int score = studentsResults[studentIndex][subjectIndex];

                if (score > bestStudentScore) {
                    bestStudentScore = score;
                    bestStudentNumber = studentIndex;
                    bestSubjectNumber = subjectIndex;
                }

                if (score < worstStudentScore) {
                    worstStudentScore = score;
                    worstStudentNumber = studentIndex;
                    worstSubjectNumber = subjectIndex;
                }
            }
        }

        overallScoreDetails[0][0] = ++bestStudentNumber;
        overallScoreDetails[0][1] = ++bestSubjectNumber;
        overallScoreDetails[0][2] = bestStudentScore;

        overallScoreDetails[1][0] = ++worstStudentNumber;
        overallScoreDetails[1][1] = ++worstSubjectNumber;
        overallScoreDetails[1][2] = worstStudentScore;

        return overallScoreDetails;
    }

    public int[][] getClassSummaryDetails() {

        int[][] classSummaryDetails = new int[2][2];

        int[] studentTotalScore = getStudentTotalScore();

        int worstStudentNumber = 0;
        int worstStudentTotalScore = studentTotalScore[0];

        int bestStudentNumber = 0;
        int bestStudentTotalScore = studentTotalScore[0];

        for (int index = 1; index < studentCount; index++) {

            int totalScore = studentTotalScore[index];

            if (totalScore < worstStudentTotalScore) {
                worstStudentTotalScore = totalScore;
                worstStudentNumber = index;
            }

            if (totalScore > bestStudentTotalScore) {
                bestStudentTotalScore = totalScore;
                bestStudentNumber = index;
            }
        }

        classSummaryDetails[0][0] = ++bestStudentNumber;
        classSummaryDetails[0][1] = bestStudentTotalScore;

        classSummaryDetails[1][0] = ++worstStudentNumber;
        classSummaryDetails[1][1] = worstStudentTotalScore;

        return classSummaryDetails;
    }

    public int getClassTotalScore() {

        int classTotalScore = 0;

        int[] studentTotalScore = getStudentTotalScore();

        for (int totalScore : studentTotalScore)
            classTotalScore += totalScore;

        return classTotalScore;
    }

    // studentsResults

    public void displayStudentScoreTable() {
        int[] position = getPosition();

        System.out.println("=====================================================");
        System.out.printf("%-15s", "STUDENT");

        for (int index = 0; index < subjectCount; index++) {
            System.out.printf("SUB%d  ", index + 1);
        }

        System.out.printf("%-5s%-7s%s%n", "TOT", "AVE", "POS");
        System.out.println("=====================================================");

        for (int index = 0; index < studentCount; index++) {
            System.out.printf("Student %-7d", index + 1);

            int scoreTotal = 0;

            for (int subjectIndex = 0; subjectIndex < subjectCount; subjectIndex++) {
                System.out.printf("%-6d", studentsResults[index][subjectIndex]);

                scoreTotal += studentsResults[index][subjectIndex];
            }

            double averageScore = (double) scoreTotal / subjectCount;
            System.out.printf("%-5d%-7.2f%d%n", scoreTotal, averageScore, position[index]);
        }

        System.out.println("=====================================================\n");
    }

    public void displaySubjectSummary() {
        System.out.println("=====================================================");
        System.out.println("SUBJECT SUMMARY");

        for (int index = 0; index < subjectCount; index++) {
            String subjectSummary = getSummaryForSubject(index);
            System.out.println(subjectSummary);
        }

        int[][] subjectDifficultyDetails = getSubjectDifficultyDetails();

        int[][] overallScoreDetails = getOverallScoreDetails();

        String otherSubjectSummary = String.format("""
        The hardest subject is Subject %d with %d failures
        The easiest subject is Subject %d with %d passes
        The overall Highest score is scored by Student %d in subject %d scoring %d
        The overall Lowest score is scored by Student %d in subject %d scoring %d
        """,
        subjectDifficultyDetails[0][0],
        subjectDifficultyDetails[0][1],
        subjectDifficultyDetails[1][0],
        subjectDifficultyDetails[1][1],
        overallScoreDetails[0][0],
        overallScoreDetails[0][1],
        overallScoreDetails[0][2],
        overallScoreDetails[1][0],
        overallScoreDetails[1][1],
        overallScoreDetails[1][2]
        );
        System.out.print(otherSubjectSummary);
        System.out.println("=====================================================\n");
    }

    public void displayClassSummary() {
        int[][] classSummaryDetails = getClassSummaryDetails();
        int classTotalScore = getClassTotalScore();
        double averageScore = (double) classTotalScore / studentCount;

        System.out.println("CLASS SUMMARY");
        System.out.println("=====================================================");
        System.out.printf("Best Graduating Student is: Student %d scoring %d%n",
            classSummaryDetails[0][0], classSummaryDetails[0][1]);
        System.out.println("=====================================================\n");

        System.out.println("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!");
        System.out.printf("Worst Graduating Student is: Student %d scoring %d%n",
            classSummaryDetails[1][0], classSummaryDetails[1][1]);
        System.out.println("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n");

        System.out.println("=====================================================");
        System.out.println("Class total score is: " + classTotalScore);
        System.out.printf("Class Average score is: %.2f%n", averageScore);
        System.out.println("=====================================================\n");
    }
}